File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/queues/specs/GivenAStandardQueue/AndASecondFileThatFailsToPrepare/WhenTheQueueIsStarted.java
package com.lasthopesoftware.bluewater.client.playback.queues.specs.GivenAStandardQueue.AndASecondFileThatFailsToPrepare;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.IPlaybackPreparer;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.FakeBufferingPlaybackHandler;
import com.lasthopesoftware.bluewater.client.playback.queues.CompletingFileQueueProvider;
import com.lasthopesoftware.bluewater.client.playback.queues.PreparedPlaybackQueue;
import com.lasthopesoftware.messenger.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeoutException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenTheQueueIsStarted {

	private static TimeoutException timeoutException;
	private static boolean firstPromiseCancelled;

	@BeforeClass
	public static void before() {
		final Random random = new Random(System.currentTimeMillis());

		final List<ServiceFile> serviceFiles =
			Stream
				.range(0, 2)
				.map(i -> new ServiceFile(random.nextInt()))
				.collect(Collectors.toList());

		final IPlaybackPreparer playbackPreparer = mock(IPlaybackPreparer.class);
		when(playbackPreparer.promisePreparedPlaybackHandler(any(), anyInt()))
			.thenReturn(new Promise<>(new FakeBufferingPlaybackHandler()))
			.thenReturn(new Promise<>(messenger -> messenger.cancellationRequested(() -> firstPromiseCancelled = true)));

		final CompletingFileQueueProvider bufferingPlaybackQueuesProvider
			= new CompletingFileQueueProvider();

		final int startPosition = 0;

		final PreparedPlaybackQueue queue = new PreparedPlaybackQueue(
			() -> 2,
			playbackPreparer,
			bufferingPlaybackQueuesProvider.provideQueue(serviceFiles, startPosition));

		queue.promiseNextPreparedPlaybackFile(0)
			.eventually(p -> queue.promiseNextPreparedPlaybackFile(0))
			.excuse(e -> {
				if (e instanceof TimeoutException)
					timeoutException = (TimeoutException)e;

				return null;
			});
	}

	@Test
	public void thenATimeoutExceptionIsThrown() {
		assertThat(timeoutException).isNotNull();
	}

	@Test
	public void thenTheFirstPromiseIsCancelled() {
		assertThat(firstPromiseCancelled).isTrue();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/GivenAStandardQueue/AndTheSecondFileThrowsAnExceptionOnFirstPreparation/WhenTheQueueIsStarted.kt
package com.lasthopesoftware.bluewater.client.playback.engine.preparation.GivenAStandardQueue.AndTheSecondFileThrowsAnExceptionOnFirstPreparation

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.access.FakePlaybackQueueConfiguration
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakeBufferingPlaybackHandler
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakePreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.joda.time.Duration
import org.junit.jupiter.api.Test

private const val libraryId = 8

class WhenTheQueueIsStarted {

	private val expectedPlaybackHandler = FakeBufferingPlaybackHandler()
	private val returnedPlaybackHandler by lazy {
		val serviceFiles = (0..2).map(::ServiceFile)

		val playbackPreparer = mockk<PlayableFilePreparationSource>().apply {
			every { promisePreparedPlaybackFile(LibraryId(libraryId), ServiceFile(0), Duration.ZERO) } returns Promise(FakePreparedPlayableFile(FakeBufferingPlaybackHandler()))
			every { promisePreparedPlaybackFile(LibraryId(libraryId), ServiceFile(1), Duration.ZERO) } returnsMany listOf(
				Promise(Exception()),
				Promise(FakePreparedPlayableFile(expectedPlaybackHandler)),
			)
		}

		val bufferingPlaybackQueuesProvider = CompletingFileQueueProvider()
		val startPosition = 0
		val queue = PreparedPlayableFileQueue(
			FakePlaybackQueueConfiguration(2),
			playbackPreparer,
			bufferingPlaybackQueuesProvider.provideQueue(LibraryId(libraryId), serviceFiles, startPosition))

		queue.promiseNextPreparedPlaybackFile(Duration.ZERO)
			?.eventually { queue.promiseNextPreparedPlaybackFile(Duration.ZERO) }
			?.toExpiringFuture()
			?.get()
			?.playableFile
	}

	@Test
	fun `then the expected playback handler is returned`() {
		assertThat(returnedPlaybackHandler).isEqualTo(expectedPlaybackHandler).isNotNull
	}
}
